internal/engine/RATIONALE.md

```diff
@@ -0,0 +1,38 @@
+# Notable rationale of wazero (Engine)
+
+## Tail Call Implementation
+
+The tail call implementation in wazero takes some liberties at interpreting the WebAssembly spec, with the goal of correctness, if not strict adherence to the spec. In principle
+
+- in the interpreter, tail calls are implemented by simply resetting the program counter to the start of the function being called
+- in the compiler, tail calls are implemented by emitting a jump instruction to the start of the function.
+
+These general rules are however subject to some exceptions, which are detailed below.
+
+### Interpreter
+
+There are 3 cases that needs to be handled in the interpreter:
+
+1. **Tail calls to the same function**: this is the simplest case, where we just reset the program counter to the start of the function body. This is straightforward and does not require any special handling.
+2. **Tail calls to a different function**: this is also straightforward, as we just reset the program counter to the start of the function body, then replace the function body with the new function's body. 
+3. **Tail calls to a host function**: in this case, we cannot simply reset the program counter, as the host function is not part of the WebAssembly module. Instead, we fall back to a plain call to the host function, which is the same as if we were calling it from a regular function.
+
+As a further restriction, currently the interpreter does not handle tail calls to any **imported functions** and falls back to a plain call in these cases too.
```

Comment:
This means functions from other modules? Are there more?

Reply:
could you describe on why part of this restriction (that's the rationale i guess? this is not a rationale)